plugins/tangle/solidifier.go

```diff
@@ -246,33 +246,135 @@ func solidQueueCheck(cachedTxMetas map[string]*tangle.CachedMetadata, milestoneI
+}
 
-		// all solidified txs are newly solidified => propagate all
-		for _, txHash := range txsToSolidify {
-			for _, approverHash := range tangle.GetApproverHashes(txHash) {
-				cachedApproverTxMeta := tangle.GetCachedTxMetadataOrNil(approverHash) // meta +1
-				if cachedApproverTxMeta == nil {
-					continue
+// solidifyFutureCone updates the solidity of the future cone (transactions approving the given transactions).
+// we have to walk the future cone, and update the solidity of all transactions by walking their past cone.
+// as a special property, invocations of the yielded function share the same 'already traversed' set to circumvent
+// walking the future cone of the same transactions multiple times.
+// all cachedTxMetas have to be released outside.
+func solidifyFutureCone(cachedTxMetas map[string]*tangle.CachedMetadata, txHashes hornet.Hashes, abortSignal chan struct{}) error {
+	traversed := map[string]struct{}{}
+	nonSolidTxs := map[string]struct{}{}
+
+	// this is an efficient way to update the whole cone, because it will not be recursive,
+	// since we check if we already traversed the future cone of the transactions we want to walk.
+	for _, txHash := range txHashes {
+
+		if err := dag.TraverseApprovers(txHash,
+			// traversal stops if no more transactions pass the given condition
+			func(cachedTxMeta *tangle.CachedMetadata) (bool, error) { // meta +1
+				defer cachedTxMeta.Release(true) // meta -1
+
+				if _, exists := cachedTxMetas[string(cachedTxMeta.GetMetadata().GetTxHash())]; !exists {
+					// release the tx metadata at the end to speed up calculation
+					cachedTxMetas[string(cachedTxMeta.GetMetadata().GetTxHash())] = cachedTxMeta.Retain()
+				}
+
+				if _, previouslyTraversed := traversed[string(cachedTxMeta.GetMetadata().GetTxHash())]; previouslyTraversed {
+					// do not walk the future cone again if we already traversed this tx before
+					return false, nil
 				}
 
-				if cachedApproverTxMeta.GetMetadata().IsSolid() {
-					// Do not propagate already solid Txs
+				// mark the tx as traversed, so we don't check the past cone again if it was not solid
+				traversed[string(cachedTxMeta.GetMetadata().GetTxHash())] = struct{}{}
 
-					// Do no force release here, otherwise cacheTime for new Tx could be ignored
-					cachedApproverTxMeta.Release() // meta -1
-					continue
+				coneSolid, err := solidifyPastCone(cachedTxMetas, nonSolidTxs, cachedTxMeta.Retain(), abortSignal)
+				if err != nil {
+					return false, err
 				}
 
-				checkSolidityAndPropagate(cachedApproverTxMeta.Retain()) // tx pass +1
+				if !coneSolid {
+					// mark the tx to have a non-solid cone
+					// this is used to reduce recursion
+					nonSolidTxs[string(cachedTxMeta.GetMetadata().GetTxHash())] = struct{}{}
+				}
+
+				// do not walk the future cone if the current past cone is already non-solid
+				return coneSolid, nil
+			},
+			// consumer
+			func(cachedTxMeta *tangle.CachedMetadata) error { // meta +1
```

Comment:
do we maybe want to make the consumer optional?

Reply:
OK, pass nil ?

Reply:
yes